analyzers/src/SonarAnalyzer.CSharp/Rules/UseCachedRegex.cs

```diff
@@ -0,0 +1,114 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class UseCachedRegex : UseCachedRegexBase<SyntaxKind>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                var objectCreationExpression = (ObjectCreationExpressionSyntax)c.Node;
+
+                if (objectCreationExpression.NameIs(nameof(Regex))
+                    && IsWithinCorrectContext(objectCreationExpression)
+                    && IsCorrectObjectType(objectCreationExpression, c.SemanticModel)
+                    && IsArgumentConstantOrReadOnly(objectCreationExpression, c.SemanticModel)
+                    && !IsCompliantAssignment(objectCreationExpression, c.SemanticModel))
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, objectCreationExpression.GetLocation()));
+                }
+            },
+            SyntaxKind.ObjectCreationExpression);
+
+    private static bool IsCompliantAssignment(ObjectCreationExpressionSyntax objectCreationExpression, SemanticModel model) =>
+        objectCreationExpression.Parent switch
+        {
+            AssignmentExpressionSyntax { RawKind: (int)SyntaxKind.SimpleAssignmentExpression } assignmentExpression =>
+                model.GetSymbolInfo(assignmentExpression.Left).Symbol is IFieldSymbol { IsReadOnly: true } or IPropertySymbol { IsReadOnly: true }
+                || IsAssignmentWithinIfStatement(assignmentExpression, model),
+            AssignmentExpressionSyntax { RawKind: (int)SyntaxKindEx.CoalesceAssignmentExpression } assignmentExpression =>
+                model.GetSymbolInfo(assignmentExpression.Left).Symbol is IFieldSymbol or IPropertySymbol,
+            BinaryExpressionSyntax
+            {
+                RawKind: (int)SyntaxKind.CoalesceExpression,
+                Parent: AssignmentExpressionSyntax { RawKind: (int)SyntaxKind.SimpleAssignmentExpression or (int)SyntaxKindEx.CoalesceAssignmentExpression } assignmentExpression
+            } coalesceExpression => coalesceExpression.Left.GetName() == assignmentExpression.Left.GetName()
```

Comment:
Why do you care if the coalesce and the assignment `left` point to the same thing?

As I understand it, you are NOT rasing for this:

```csharp
someProperty = someProperty ?? new Regex("pattern");
```

...but want to raise for this?

```csharp
Regex x = null;
someProperty = x ?? new Regex("pattern");
```

Is there a reason these two cases are different for you, or am I missing something?
Also, you can have the reverse, were the `new Regex()` node is on the `left` of the `??` operator (even though it's a bad practise anyways)

Reply:
The second case is a good question. This might produce FPs, but this also could lead to multiple `Regex` instantiation. Not sure what would be the best approach here.

The reverse should raise as the `??` operator will always evaluate left as not `null` and therefore always create a new instance. So it is not a compliant assignment.